goth/assertions/assertions.py

```diff
@@ -78,10 +78,21 @@ class Assertion(AsyncIterable[E]):
     _generator: Optional[AsyncIterator[E]]
     """An asynchronous generator that provides events to the assertion coroutine."""
 
-    def __init__(self, events: Sequence[E], func: AssertionFunction) -> None:
+    def __init__(
+        self, events: Sequence[E], func: AssertionFunction, name: Optional[str] = None
+    ) -> None:
+        """Create an assertion that processes `events` as prescribed by `func`.
+
+        The `name` argument may be `None` if `func` refers to a function defined with
+        `(async) def func(...)`. In such cases, the assertion's name will be constructed
+        from the `__module__` and `__qualname__` attributes of `func`.
+        Otherwise `name` has to be provided or an exception will be raised.
```

Comment:
What exception would be raised in this case?

Reply:
Currently it would be `AttributeError` but this is an implementation detail. What I was trying to say is that if `func` is not defined by `async def` and `name` is not specified then the behaviour is undefined, an exception may be thrown or some other terrible thing may happen.

Reply:
Got it! Then perhaps my other comment on the test file is invalid, I'll leave it up to you to decide whether it makes sense to add another test case.

Reply:
I've decided to raise `ValueError` when the name cannot be constructed and updated the docstring accordingly, but haven't added another test case.